crates/ast/src/lib.rs

```diff
@@ -1,33 +1,93 @@
 //! source: https://kotlinlang.org/spec/syntax-and-grammar.html#syntax-grammar
-
 mod nodes;
+use nodes::*;
+use serde::Serialize;
+use syntax::SyntaxNode;
+
+#[macro_export]
+macro_rules! ast_node {
+    ($ast:ident, $kind:ident) => {
+        #[derive(PartialEq, Eq, Hash, Clone)]
+        #[repr(transparent)]
+        pub struct $ast(::syntax::SyntaxNode);
+        impl $ast {
+            #[allow(unused)]
+            pub fn cast(node: ::syntax::SyntaxNode) -> Option<Self> {
+                if node.kind() == ::syntax::SyntaxKind::Syntax($kind) {
+                    Some(Self(node))
+                } else {
+                    None
+                }
+            }
+        }
+    };
+}
```

Comment:
This would not work, using a a proc macro for this instead. See macros/src/parse.rs